### PR TITLE
Add retry when sending record to clickhouse in theia_clickhouse_test

### PR DESCRIPTION
**Pull request by yuntanghsu**
For some reason, ClickHouse on jenkins-e2e-test will receive termination signal, which results in the disconnection with zookeeper. It takes time (~30s) for clickhouse to fix the error `error: Couldn't start replication (table will be in readonly mode)` by itself.

Adding retry to commit the query later.

Signed-off-by: Yun-Tang Hsu <hsuy@vmware.com>

### CODE REVIEW

**Comment by yanjunz97** [comment]
May I learn about why do you choose to increase the ZK replica number instead of decrease the commit interval? It is a little bit strange to me that we need to deploy 3 ZK replicas to support only 1 ClickHouse replica. And I also notice the flow visibility e2e test is always stable which may indicate single-replica ZK can handle the flows when insertion rate is under some value? BTW, I'm not sure if increasing the ZK memory size would help. The default value is 5GB, but I think we can increase this default value if it helps.

**Comment by yuntanghsu** [comment]
> May I learn about why do you choose to increase the ZK replica number instead of increase the commit interval? It is a little bit strange to me that we need to deploy 3 ZK replicas to support only 1 ClickHouse replica. And I also notice the flow visibility e2e test is always stable which may indicate single-replica ZK can handle the flows when insertion rate is under some value? BTW, I'm not sure if increasing the ZK memory size would help. The default value is 5GB, but I think we can increase this default value if it helps.

The problem is that, for some reason, jenkins-e2e test is not passed every time. And I notice that, it might because of the limited resource on jenkins' node. So the zookeeper will keep disconnecting when we run the test and making the table in readonly mode. 

`Error:      	Received unexpected error:
        	            	code: 242, message: Table is in readonly mode (zookeeper path: /clickhouse/tables/0/default/.inner.flows_node_view_local)`

File: test/e2e/theia_clickhouse_test.go [hunk 1]
```diff
@@ -90,12 +91,13 @@ const (
...
 	recordPerCommit      = 1000
 	insertInterval       = 1
-	threshold            = 25
+	threshold            = 35
 	MaxInt32             = 1<<31 - 1
 	numFieldsInDiskInfo  = 9
 	numFieldsInTableInfo = 7
 	dateBaseName         = "default"
 	defaultPath          = "/var/lib/clickhouse/"
+	flowsRelatedTableNum = 4
```
Comment (dreamtalen) [inline test/e2e/theia_clickhouse_test.go:100]: May I ask which four tables are related?

File: test/e2e/theia_clickhouse_test.go [hunk 2]
```diff
@@ -270,8 +272,8 @@ func testTheiaGetClickHouseInsertRate(t *testing.T, data *TestData, connect *sql
 		assert.Equal(4, len(tableInfoArray), "tableInfoArray: %s", tableInfoArray)
 		actualInsertRate, error := strconv.Atoi(tableInfoArray[1])
 		assert.NoError(error)
-		tableNum := len(tableColumnNumberMap)
-		percent := (actualInsertRate/tableNum - recordPerCommit/insertInterval) * 100 / (recordPerCommit / insertInterval)
+		percent := (actualInsertRate/flowsRelatedTableNum - recordPerCommit/insertInterval) * 100 / (recordPerCommit / insertInterval)
+		percent = int(math.Abs(float64(percent)))
 		assert.LessOrEqualf(percent, threshold, "stdout: %s, expectedInsertRate: %s", stdout, recordPerCommit/insertInterval)
```
Comment (dreamtalen) [inline test/e2e/theia_clickhouse_test.go:277]: I think these 3 lines of code could be simplified using `assert.InDeltaf`.
Reference: we are comparing the throughput difference should be less than 15% in https://github.com/antrea-io/theia/blob/main/test/e2e/flowvisibility_test.go#L760

**Comment by yanjunz97** [comment]
> > May I learn about why do you choose to increase the ZK replica number instead of increase the commit interval? It is a little bit strange to me that we need to deploy 3 ZK replicas to support only 1 ClickHouse replica. And I also notice the flow visibility e2e test is always stable which may indicate single-replica ZK can handle the flows when insertion rate is under some value? BTW, I'm not sure if increasing the ZK memory size would help. The default value is 5GB, but I think we can increase this default value if it helps.
> 
> The problem is that, for some reason, jenkins-e2e test is not passed every time. And I notice that, it might because of the limited resource on jenkins' node. So the zookeeper will keep disconnecting when we run the test and making the table in readonly mode.
> 
> `Error: Received unexpected error: code: 242, message: Table is in readonly mode (zookeeper path: /clickhouse/tables/0/default/.inner.flows_node_view_local)`

Do you mean even if the insertion rate is low, we still get this error in ClickHouse CLI test? I'm just curious about why we do not see the similar failure in TestPolicyRecommendation or TestFlowVisibility, as I think these two tests should consume more resources than TestTheiaClickHouseStatusCommand. Both of these tests deploy extra components like Grafana and Spark, but the ZooKeeper has not been disconnected in these two tests.

File: test/e2e/theia_clickhouse_test.go (hunk 1 above)
Comment (yuntanghsu) [reply test/e2e/theia_clickhouse_test.go:100]: inner.flows_node_view_local, inner.flows_pod_view_local, inner.flows_policy_view_loca, flows_local

**Comment by yuntanghsu** [comment]
> Do you mean even if the insertion rate is low, we still get this error in ClickHouse CLI test? I'm just curious about why we do not see the similar failure in TestPolicyRecommendation or TestFlowVisibility, as I think these two tests should consume more resources than TestTheiaClickHouseStatusCommand. Both of these tests deploy extra components like Grafana and Spark, but the ZooKeeper has not been disconnected in these two tests.

I tried to use {recordPerCommit=4000, insertInterval=8} and still saw the issue sometimes. And the reason I used such high recordPerCommit is because the insertionRate in clickhouse is for all tables (not only for the 4 flows tables). It will have around 160 rows/s even there is not traffic sent.

File: test/e2e/theia_clickhouse_test.go (hunk 2 above)
Comment (yuntanghsu) [reply test/e2e/theia_clickhouse_test.go]: Thanks for the suggestion.

File: test/e2e/theia_clickhouse_test.go [hunk 3]
```diff
@@ -350,23 +345,35 @@ func addFakeRecord(t *testing.T, stmt *sql.Stmt) {
...
-	tx, err := connect.Begin()
-	require.NoError(t, err)
-	stmt, _ := tx.Prepare(insertQuery)
-	defer stmt.Close()
-	for j := 0; j < recordPerCommit; j++ {
-		addFakeRecord(t, stmt)
-	}
-	err = tx.Commit()
-	assert.NoError(t, err)
+	err := wait.PollImmediate(5*time.Second, 90*time.Second, func() (bool, error) {
```
Comment (dreamtalen) [inline test/e2e/theia_clickhouse_test.go:348]: Better naming some constants for `5*time.Second` and `90*time.Second`

**Review by dreamtalen — APPROVED** [review]
overall LGTM

**Review by yanjunz97 — APPROVED** [review]
LGTM

**Pull request merged**